### ISSUE TITLE
[question] how conan resolves shared library dependencies

### ISSUE THREAD

**Issue by sigidagi** [conan-io/conan#16085]
Hi,
conanfile.txt
```
[requires]
libdwarf/20191104
libelf/0.8.13
[options]
libdwarf/*:shared = True
libelf/*:shared = True
[generators]
CMakeDeps
CMakeToolchain
[layout]
cmake_layout
```
after installing with command `conan install . --build=missing` I can see that it generates two `libdwarf` shared libraries:
```
sigis:dwarfas$ find ~/.conan2 -name libdwarf.so
/home/sigis/.conan2/p/b/libdwa475130d18fa7/b/build/Debug/libdwarf/libdwarf.so
/home/sigis/.conan2/p/b/libdwa475130d18fa7/p/lib/libdwarf.so
```

where in one dependency tree as expected:
```
sigis:dwarfas$ ldd /home/sigis/.conan2/p/b/libdwa475130d18fa7/b/build/Debug/libdwarf/libdwarf.so
        linux-vdso.so.1 (0x00007ffdd7aab000)
        libelf.so.0 => /home/sigis/.conan2/p/b/libele54df64ea906f/p/lib/libelf.so.0 (0x00007fd36db80000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd36d800000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fd36dc5d000)
  ```
  
  in second one - `libelf` not found:
  ```
  sigis:dwarfas$ ldd /home/sigis/.conan2/p/b/libdwa475130d18fa7/p/lib/libdwarf.so
        linux-vdso.so.1 (0x00007fffc4180000)
        **libelf.so.0 => not found**
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f73bf000000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f73bf3af000)
  ```
  
  What explanation  could be? 
  My application is linking to the last one shared library.  
 

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by jcar87** [comment]
Hi @sigidagi, thanks for opening this issue.

When you call `conan install`, the generators (in your case `CMakeDeps` and `CMakeToolchain`) contain all the information needed to tell CMake (which in turn, tells the linker) where to find the shared library dependencies from Conan. Then CMake "embeds" the location of these libraries in the RPATH of the linked binary targets (executables and shared libraries) of your project, so when you run your executables it should just work (although there are some cases related with indirect dependencies where this may not be the case, if you are experiencing this please let us know).

In addition to the `CMakeDeps` and `CMakeToolchain` generators, Conan 2.x will also automatically invoke the `VirtualRunEnv` and `VirtualBuildEnv` generators. In particular, `VirtualRunEnv` will set `LD_LIBRARY_PATH` such that the runtime linker can locate the correct shared library dependencies.

In your case, invoking `ldd` _without_ activating `VirtualRunEnv` first will cause the linker to either fail to locate libraries, or to incorrectly locate the system-managed versions. 

In summary:
* The build integrations help locate shared libraries at build time, and CMake embedding the rpaths in the linked executables also resolve the situation at runtime
* `ldd` should not be called in isolation against specific linked binaries inside the Conan cache - you must `source conanrun.sh` such that `LD_LIBRARY_PATH` is correctly defined.

**Comment by sigidagi** [comment]
Thanks @jcar87 for your quick answer. 
Running `source ./build/Debug/generators/conanrun.sh` makes dependencies in place.
